## 1

In [1]:
library("openxlsx")
library("dplyr")

Warning message:
"package 'openxlsx' was built under R version 3.6.3"Warning message:
"package 'dplyr' was built under R version 3.6.3"
Attaching package: 'dplyr'

The following objects are masked from 'package:stats':

    filter, lag

The following objects are masked from 'package:base':

    intersect, setdiff, setequal, union



In [2]:
df <- read.xlsx("./data_input/GAZ.xlsx")
head(df)

дата.замера Ртр.МПа Туст.°С газ.м3/сут конд-т.м3/сут вода.м3/сут ID  Куст
1 42343       15.5            771912                               110 К-А 
2 13/05/2015  150     73.4    827421                               110 К-А 
3 16/05/2015  16.7    71      443574                               110 К-А 
4 20/05/2015  17.9    66      304479                               110 К-А 
5 26/05/2015  17.5    70      361939     2.72          2.44        110 К-А 
6 42191       176     70      349748     3.6           1.49        110 К-А 
  Группа     
1 Кувачи-Алат
2 Кувачи-Алат
3 Кувачи-Алат
4 Кувачи-Алат
5 Кувачи-Алат
6 Кувачи-Алат

Полям загруженных данных даём чуть более понятные названия

In [3]:
colnames(df) <- c("Дата замера", "Давление", "Температура", "Добыча", "Конденсат", "Вода", "ID", "Куст", "Группа")
head(df)

Дата замера Давление Температура Добыча Конденсат Вода ID  Куст Группа     
1 42343       15.5                 771912                110 К-А  Кувачи-Алат
2 13/05/2015  150      73.4        827421                110 К-А  Кувачи-Алат
3 16/05/2015  16.7     71          443574                110 К-А  Кувачи-Алат
4 20/05/2015  17.9     66          304479                110 К-А  Кувачи-Алат
5 26/05/2015  17.5     70          361939 2.72      2.44 110 К-А  Кувачи-Алат
6 42191       176      70          349748 3.6       1.49 110 К-А  Кувачи-Алат

Дальше идёт преобразование разных форматов даты в человеческий. Если с форматом *d/m/Y* всё легко, то с датами по типу *42343* приходится использовать дополнительные библиотеки. Но всё равно каким-то образом в глубинах таблички создаются **NA**...

In [4]:
df[1] <- sapply(df[1], function(x) { ifelse(is.na(as.numeric(x)), as.character(as.Date(x, format="%d/%m/%Y")), as.character(convertToDate(x, origin = "1900-01-01"))) })
head(df)

Warning message in ifelse(is.na(as.numeric(x)), as.character(as.Date(x, format = "%d/%m/%Y")), :
"в результате преобразования созданы NA"Warning message in convertToDate(x, origin = "1900-01-01"):
"в результате преобразования созданы NA"

Дата замера Давление Температура Добыча Конденсат Вода ID  Куст Группа     
1 2015-12-05  15.5                 771912                110 К-А  Кувачи-Алат
2 2015-05-13  150      73.4        827421                110 К-А  Кувачи-Алат
3 2015-05-16  16.7     71          443574                110 К-А  Кувачи-Алат
4 2015-05-20  17.9     66          304479                110 К-А  Кувачи-Алат
5 2015-05-26  17.5     70          361939 2.72      2.44 110 К-А  Кувачи-Алат
6 2015-07-06  176      70          349748 3.6       1.49 110 К-А  Кувачи-Алат

In [5]:
typeof(df$ID)

[1] "character"

Далее и до конца пункта идёт преобразование типов полей в нужные

In [6]:
df$ID <- factor(df$ID)
df$Куст <- factor(df$Куст)
df$Группа <- factor(df$Группа)

In [7]:
head(df)
tail(df)

Дата замера Давление Температура Добыча Конденсат Вода ID  Куст Группа     
1 2015-12-05  15.5                 771912                110 К-А  Кувачи-Алат
2 2015-05-13  150      73.4        827421                110 К-А  Кувачи-Алат
3 2015-05-16  16.7     71          443574                110 К-А  Кувачи-Алат
4 2015-05-20  17.9     66          304479                110 К-А  Кувачи-Алат
5 2015-05-26  17.5     70          361939 2.72      2.44 110 К-А  Кувачи-Алат
6 2015-07-06  176      70          349748 3.6       1.49 110 К-А  Кувачи-Алат

Дата замера Давление Температура Добыча Конденсат Вода ID   Куст
28157 2019-03-25  2.5      50          125700 0.4       2    1043 КП 2
28158 2019-04-18  2.6      55          125760 0.4       2.1  1043 КП 2
28159 2019-04-19  2.6      55          125500 0.4       3    1043 КП 2
28160 2019-11-05  2.6      55          124100 0.3       2.8  1043 КП 2
28161 2019-05-22  2.6      56          123902 0         2.98 1043 КП 2
28162 2019-05-23  2.5      56          123786 0.26      2.8  1043 КП 2
      Группа       
28157 Западные Шады
28158 Западные Шады
28159 Западные Шады
28160 Западные Шады
28161 Западные Шады
28162 Западные Шады

In [8]:
df <- transform(df, Давление = as.numeric(Давление))
df <- transform(df, Температура = as.numeric(Температура))
df <- transform(df, Добыча = as.numeric(Добыча))
df <- transform(df, Конденсат = as.numeric(Конденсат))
df <- transform(df, Вода = as.numeric(Вода))
head(df)

Дата.замера Давление Температура Добыча Конденсат Вода ID  Куст Группа     
1 2015-12-05   15.5      NA        771912   NA        NA 110 К-А  Кувачи-Алат
2 2015-05-13  150.0    73.4        827421   NA        NA 110 К-А  Кувачи-Алат
3 2015-05-16   16.7    71.0        443574   NA        NA 110 К-А  Кувачи-Алат
4 2015-05-20   17.9    66.0        304479   NA        NA 110 К-А  Кувачи-Алат
5 2015-05-26   17.5    70.0        361939 2.72      2.44 110 К-А  Кувачи-Алат
6 2015-07-06  176.0    70.0        349748 3.60      1.49 110 К-А  Кувачи-Алат

## 2

Очистка от строк с пропущенными значениями

In [9]:
df <- df[rowSums(is.na(df)) == 0,]
head(df)

Дата.замера Давление Температура Добыча Конденсат Вода ID  Куст Группа     
5  2015-05-26   17.5    70          361939 2.72      2.44 110 К-А  Кувачи-Алат
6  2015-07-06  176.0    70          349748 3.60      1.49 110 К-А  Кувачи-Алат
7  2015-06-14   17.3    69          310185 3.26      1.43 110 К-А  Кувачи-Алат
8  2015-06-20   17.5    69          265176 2.97      0.25 110 К-А  Кувачи-Алат
9  2015-06-26   17.4    68          263621 3.00      1.80 110 К-А  Кувачи-Алат
10 2015-02-07   16.0    73          446708 3.49      1.87 110 К-А  Кувачи-Алат

## 3

Преобразование температуры из шкалы Цельсия в Кельвины

In [10]:
df <- transform(df, Температура = Температура + 273.15)
head(df)

Дата.замера Давление Температура Добыча Конденсат Вода ID  Куст Группа     
5  2015-05-26   17.5    343.15      361939 2.72      2.44 110 К-А  Кувачи-Алат
6  2015-07-06  176.0    343.15      349748 3.60      1.49 110 К-А  Кувачи-Алат
7  2015-06-14   17.3    342.15      310185 3.26      1.43 110 К-А  Кувачи-Алат
8  2015-06-20   17.5    342.15      265176 2.97      0.25 110 К-А  Кувачи-Алат
9  2015-06-26   17.4    341.15      263621 3.00      1.80 110 К-А  Кувачи-Алат
10 2015-02-07   16.0    346.15      446708 3.49      1.87 110 К-А  Кувачи-Алат

## 5

В этом пункте создаются безразмерные поля по заданию:

* Отношение добычи газа к конденсату;
* Отношение добычи газа к добыче воды;
* Отношение добычи воды к добыче конденсата.

При их создании в некоторых случаях образуются значения **NA** и **Inf**, что будет учтено далее

!! Alert: Inf

In [11]:
df["НефтиНаКонденсат"] <- df["Добыча"] / df["Конденсат"]
head(df)

Дата.замера Давление Температура Добыча Конденсат Вода ID  Куст Группа     
5  2015-05-26   17.5    343.15      361939 2.72      2.44 110 К-А  Кувачи-Алат
6  2015-07-06  176.0    343.15      349748 3.60      1.49 110 К-А  Кувачи-Алат
7  2015-06-14   17.3    342.15      310185 3.26      1.43 110 К-А  Кувачи-Алат
8  2015-06-20   17.5    342.15      265176 2.97      0.25 110 К-А  Кувачи-Алат
9  2015-06-26   17.4    341.15      263621 3.00      1.80 110 К-А  Кувачи-Алат
10 2015-02-07   16.0    346.15      446708 3.49      1.87 110 К-А  Кувачи-Алат
   НефтиНаКонденсат
5  133065.81       
6   97152.22       
7   95148.77       
8   89284.85       
9   87873.67       
10 127996.56

!! Alert: Inf

In [12]:
df["НефтиНаВоду"] <- df["Добыча"] / df["Вода"]
head(df)

Дата.замера Давление Температура Добыча Конденсат Вода ID  Куст Группа     
5  2015-05-26   17.5    343.15      361939 2.72      2.44 110 К-А  Кувачи-Алат
6  2015-07-06  176.0    343.15      349748 3.60      1.49 110 К-А  Кувачи-Алат
7  2015-06-14   17.3    342.15      310185 3.26      1.43 110 К-А  Кувачи-Алат
8  2015-06-20   17.5    342.15      265176 2.97      0.25 110 К-А  Кувачи-Алат
9  2015-06-26   17.4    341.15      263621 3.00      1.80 110 К-А  Кувачи-Алат
10 2015-02-07   16.0    346.15      446708 3.49      1.87 110 К-А  Кувачи-Алат
   НефтиНаКонденсат НефтиНаВоду
5  133065.81         148335.7  
6   97152.22         234730.2  
7   95148.77         216912.6  
8   89284.85        1060704.0  
9   87873.67         146456.1  
10 127996.56         238881.3

!! Alert: Inf or NA

In [13]:
df["ВодыНаКонденсат"] <- df["Вода"] / df["Конденсат"]
head(df)

Дата.замера Давление Температура Добыча Конденсат Вода ID  Куст Группа     
5  2015-05-26   17.5    343.15      361939 2.72      2.44 110 К-А  Кувачи-Алат
6  2015-07-06  176.0    343.15      349748 3.60      1.49 110 К-А  Кувачи-Алат
7  2015-06-14   17.3    342.15      310185 3.26      1.43 110 К-А  Кувачи-Алат
8  2015-06-20   17.5    342.15      265176 2.97      0.25 110 К-А  Кувачи-Алат
9  2015-06-26   17.4    341.15      263621 3.00      1.80 110 К-А  Кувачи-Алат
10 2015-02-07   16.0    346.15      446708 3.49      1.87 110 К-А  Кувачи-Алат
   НефтиНаКонденсат НефтиНаВоду ВодыНаКонденсат
5  133065.81         148335.7   0.89705882     
6   97152.22         234730.2   0.41388889     
7   95148.77         216912.6   0.43865031     
8   89284.85        1060704.0   0.08417508     
9   87873.67         146456.1   0.60000000     
10 127996.56         238881.3   0.53581662

## 6

In [14]:
getYear <- function(dateString) {
    return (as.numeric(format(as.Date(dateString),'%Y')))
}

In [15]:
df_2018 <- filter(df, getYear(Дата.замера) == 2018)
head(df_2018)

Дата.замера Давление Температура Добыча Конденсат Вода  ID  Куст Группа     
1 2018-01-01  6.9      333.75      131080 0.60      1.150 110 К-А  Кувачи-Алат
2 2018-11-01  6.7      334.15      157493 0.23      2.280 110 К-А  Кувачи-Алат
3 2018-01-22  6.7      334.15      153157 0.23      2.300 110 К-А  Кувачи-Алат
4 2018-01-27  6.7      332.15      149956 0.21      0.842 110 К-А  Кувачи-Алат
5 2018-01-28  6.8      333.65      148368 0.28      0.890 110 К-А  Кувачи-Алат
6 2018-01-02  6.8      333.15      148374 0.12      0.150 110 К-А  Кувачи-Алат
  НефтиНаКонденсат НефтиНаВоду ВодыНаКонденсат
1  218466.7        113982.61    1.916667      
2  684752.2         69075.88    9.913043      
3  665900.0         66590.00   10.000000      
4  714076.2        178095.01    4.009524      
5  529885.7        166705.62    3.178571      
6 1236450.0        989160.00    1.250000

In [16]:
mining <- aggregate(Добыча ~ Куст, df_2018, sum)
mining

Куст   Добыча  
1  К-А    34443890
2  КП 1   59002366
3  КП 11  43679297
4  КП 12  41142008
5  КП 14  23521355
6  КП 16  74119642
7  КП 19  27745838
8  КП 2   43930136
9  КП 242 53816662
10 КП 3   18471218
11 КП 301 19191073
12 КП 9   25397305
13 СШ     55608131

## 7

In [17]:
id110 <- filter(df, ID == 110)
head(id110)

Дата.замера Давление Температура Добыча Конденсат Вода ID  Куст Группа     
1 2015-05-26   17.5    343.15      361939 2.72      2.44 110 К-А  Кувачи-Алат
2 2015-07-06  176.0    343.15      349748 3.60      1.49 110 К-А  Кувачи-Алат
3 2015-06-14   17.3    342.15      310185 3.26      1.43 110 К-А  Кувачи-Алат
4 2015-06-20   17.5    342.15      265176 2.97      0.25 110 К-А  Кувачи-Алат
5 2015-06-26   17.4    341.15      263621 3.00      1.80 110 К-А  Кувачи-Алат
6 2015-02-07   16.0    346.15      446708 3.49      1.87 110 К-А  Кувачи-Алат
  НефтиНаКонденсат НефтиНаВоду ВодыНаКонденсат
1 133065.81         148335.7   0.89705882     
2  97152.22         234730.2   0.41388889     
3  95148.77         216912.6   0.43865031     
4  89284.85        1060704.0   0.08417508     
5  87873.67         146456.1   0.60000000     
6 127996.56         238881.3   0.53581662

## 8

In [18]:
filter(aggregate(Вода ~ ID, df, max), Вода <= 2)["ID"][[1]]

[1] 1010 1024 103  107  109  118  165  7   
81 Levels: 100 1008 1009 101 1010 1011 1012 1013 1014 1015 1016 1017 ... 7

## 9

In [19]:
temp <- aggregate(Добыча ~ ID, df, min)

In [20]:
all_ids <- unique(df["ID"])[[1]]

In [21]:
length(temp[[1]]) == length(all_ids) || length(temp[[1]]) == 0

[1] TRUE

Условию "*Скважины, суммарная добыча в день в которых не опускалась ниже 1000 м3/сут*" не удовлетворяет ни одна, поэтому ищем такое значение, ниже которого не опускалось лишь на 3х кустах

In [22]:
temp[order(temp$Добыча, decreasing=TRUE),][3,"Добыча"]

[1] 255457

## 10

In [23]:
groups <- aggregate(Добыча ~ Группа, df_2018, sum)
groups

Группа        Добыча   
1 Западные Шады  70176357
2 Кувачи-Алат    85671253
3 Северные Шады  55608131
4 Хаузак        308613180

Самая результативная по добыче газа группа

In [24]:
groups[order(groups$Добыча, decreasing=TRUE), ][1, "Группа"]

[1] Хаузак
Levels: Западные Шады Кувачи-Алат Северные Шады Хаузак

## 11

In [25]:
kusts <- aggregate(Вода ~ Куст, df_2018, sum)
kusts

Куст   Вода    
1  К-А     284.788
2  КП 1   1494.829
3  КП 11   319.241
4  КП 12   389.726
5  КП 14   180.844
6  КП 16   490.239
7  КП 19   374.107
8  КП 2    632.416
9  КП 242  191.167
10 КП 3    114.592
11 КП 301  126.208
12 КП 9    256.456
13 СШ      585.523

Самый результативный по добыче воды куст

In [26]:
kusts[order(kusts$Вода, decreasing=TRUE), ][1, "Куст"]

[1] КП 1
13 Levels: К-А КП 1 КП 11 КП 12 КП 14 КП 16 КП 19 КП 2 КП 242 КП 3 ... СШ

## 12

In [27]:
right_results <- filter(df, df$НефтиНаВоду != Inf)
kusts <- aggregate(НефтиНаВоду ~ Куст, right_results, mean)
kusts

Куст   НефтиНаВоду
1  К-А    1100133.1  
2  КП 1   1901441.3  
3  КП 11   463532.7  
4  КП 12   267186.6  
5  КП 14   870718.5  
6  КП 16   334753.3  
7  КП 19   416871.1  
8  КП 2    281412.9  
9  КП 242 5450558.8  
10 КП 3    178675.8  
11 КП 301 1483464.5  
12 КП 9    321463.5  
13 СШ      298763.7

Название куста, в котором среднее отношение добычи нефти к добыче воды было самым наибольшим за всё время наблюдений

In [28]:
kusts[order(kusts$НефтиНаВоду, decreasing=TRUE), ][1, "Куст"]

[1] КП 242
13 Levels: К-А КП 1 КП 11 КП 12 КП 14 КП 16 КП 19 КП 2 КП 242 КП 3 ... СШ